### PR TITLE
Adding simple simulations + method signatures

### PR DESCRIPTION
**Pull request by qwinters**

### CODE REVIEW

File: dkdisc/utils/brownianmotion.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+              increments (useful for testing).
+
+    Return:
+        randomized_steps - a collection of matrices in the form of an
+                           npt.NDArray with dimensions N x d x S where
+                            N - num_particles
+                            d - space_dim
+                            S - num_steps
+    """
+    std_dev = time_step**2
```
Comment (FCornalba) [inline dkdisc/utils/brownianmotion.py:32]: this should be std_dev = \sqrt{time_step}

File: dkdisc/utils/brownianmotion.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    randomized_steps = rng.normal(
+        scale=std_dev, size=(num_particles, space_dim, num_steps)
+    )
+    return randomized_steps
+
+
+def simulate_brownian_motion_paths(
```
Comment (FCornalba) [inline dkdisc/utils/brownianmotion.py:39]: ok, I am not sure we will need to compute the full paths (increments should suffice for all our purposes), but let's keep it anyway

File: dkdisc/dkdisc.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import scipy.ndimage as spi
+import numpy as np
+import numpy.typing as npt
+
+
+def generate_uniform_grid(mesh_size: float, space_dim: int) -> npt.NDArray:
+    assert mesh_size < 0.5, "Need a reasonable mesh size"
+    number_of_grid_points = 1 + int(1.0 / mesh_size)
+    grid_shape = tuple(number_of_grid_points for _ in range(space_dim))
+    return np.zeros(grid_shape)
```
Comment (FCornalba) [inline dkdisc/dkdisc.py:13]: this is a null function over the grid, not the grid itself

File: dkdisc/dkdisc.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    def dean_kawasaki_drift(func_values):
+        return spi.laplace(
+            func_values, mode="wrap"
+        )  # WRAP = assumption that we dealing with DK on the torus [-pi,pi)^d
+
+    def dean_kawasaki_diff_matrix(func_values, dWiener):
+        pos_func_values = np.maximum(func_values, 0)
+        sqrt_pos_func_values = np.sqrt(pos_func_values)
+        vec_field = np.multiply(sqrt_pos_func_values, dWiener)
+        discrete_divergence = np.sum(np.gradient(vec_field))
```
Comment (FCornalba) [inline dkdisc/dkdisc.py:31]: I need to double check this for myself

**Review by FCornalba — COMMENTED** [review]
minor comments so far, let's go through `dkdisc/finitediff/linear_sdesolver.py` later!

**Review by FCornalba — COMMENTED** [review]
I have reviewed everything except the experiment setup / plotting!

File: dkdisc/finitediff/__init__.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+                self.identity = np.eye(self.num_points_1d)
+            else:
+                tensor_dimension = (self.num_points_1d,) * self.dimension
+                diagonal = tuple([np.arange(self.num_points_1d)] * self.dimension)
+                identity_tensor = np.zeros(tensor_dimension)
+                identity_tensor[diagonal] = 1.0
+                self.identity = identity_tensor
+        return self.identity
+
+    def get_laplacian(self) -> npt.NDArray:
```
[the remark this answers is not in the document]
Comment (qwinters) [reply dkdisc/finitediff/__init__.py:54]: Agree!

File: dkdisc/finitediff/__init__.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+            bm_paths = np.cumsum(np.stack(bm_paths_to_keep))
+        else:
+            bm_paths = np.zeros((1, 1))
+
+        log.debug("Completed finite difference algorithm!")
+
+        return densities, bm_paths
+
+    @abc.abstractmethod
+    def first_step(
```
[the remark this answers is not in the document]
Comment (qwinters) [reply dkdisc/finitediff/__init__.py:220]: yep agree -- this is the BDF2 impl from SDELab, we can always add other solver algorithms in the future

File: dkdisc/finitediff/__init__.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+
+        # TODO: Modify API to include to option to accumulate weak soln's
+        densities_to_keep = []
+        bm_paths_to_keep = []
+        curr_time = 0.0
+
+        prev_density = initial_density
+        dWiener = self.rng.normal(
+            scale=np.sqrt(self.time_step), size=self.grid.shape
+        )
```
[the remark this answers is not in the document]
Comment (qwinters) [reply dkdisc/finitediff/__init__.py:143]: Yes, but this is actually helpful to generate them iteratively for space concern reasons (e.g. each dWiener will be garbage collected unless we explicitly ask to save it). I think the time saving will be marginal to generate them all at once

File: dkdisc/finitediff/__init__.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+import dataclasses
+import logging as log
+import numpy as np
+import numpy.typing as npt
+import scipy.linalg as lin
+import typing as typ
+
+
+@dataclasses.dataclass
+class Grid:
```
[the remark this answers is not in the document]
Comment (qwinters) [reply dkdisc/finitediff/__init__.py:11]: yep!

File: dkdisc/finitediff/linear.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        initial_density: npt.NDArray,
+        initial_noise: npt.NDArray,
+    ) -> npt.NDArray:
+
+        deterministic_vector = (
+            (1.0 - self.params.first_step_alpha)
+            * self.params.time_step
+            * self.params.drift_func(initial_density)
+        )
+        step_vector = initial_density + deterministic_vector + initial_noise
```
[the remark this answers is not in the document]
Comment (qwinters) [reply dkdisc/finitediff/linear.py:56]: I can change names / add comments if that's helpful!

File: dkdisc/finitediff/linear.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+        if self.grid.dimension > 1:
+            log.debug("Solving first step with np.linalg.tensorsolve")
+            return np.linalg.tensorsolve(self.params.first_step_matrix, step_vector)
```
[the remark this answers is not in the document]
Comment (qwinters) [reply dkdisc/finitediff/linear.py:60]: For tensor solving, yes

File: dkdisc/finitediff/__init__.py [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+
+        if keep_all_steps:
+            densities = np.stack(densities_to_keep)
+        else:
+            densities = curr_density
+
+        if keep_bm_paths:
+            bm_paths = np.cumsum(np.stack(bm_paths_to_keep))
+        else:
+            bm_paths = np.zeros((1, 1))
```
[the remark this answers is not in the document]
Comment (qwinters) [reply dkdisc/finitediff/__init__.py:213]: Can do, but I've been using an auto-formatter for things which is how I got these (see: https://black.readthedocs.io/en/stable/). 

I'm ok with whatever format / thing we want to do, it's just a little easier to be able to write super messy code and then run a CLI to do all the formatting for you

**Pull request merged**